### PR TITLE
Support non-UTF8 encodings

### PR DESCRIPTION
**Pull request by apetresc**
Status
=====
 - [x] Implement support for reading non-UTF-8 files
 - [x] Add tests for Chinese SGFs
 - [x] Add support for saving files back to their *non*-UTF-8 forms (if this is even desirable).
 - [x] More tests with more languages/encodings, and some unsupported encodings as well.

Screenshots
=========
screenshot 2016-08-24 23 47 47
(The exciting part of this screenshot is not that Chinese characters are being displayed correctly, since Sabaki can already do that; the exciting part is that the file being opened is not UTF-8 encoded, as can be seen from the terminal).

Known Issues
==========
 - [x] Currently, if the `CA[]` property comes *after* a text node in the root gametree, the text node preceding it won't be properly decoded if it's non-ISO.

Addresses yishn/sabaki#121.

*(Note: this is my first time writing any Node/Electron at all, so this may not be very idiomatic 😄)*

### CODE REVIEW

**Comment by apetresc** [comment]
On further reflection, I think the only safe way to set the encoding is to make two passes: one where we we just parse every field as ISO-8859-1 in order to read the `CA` attribute (if present), and then if it is set, another parsing round where text fields are read properly.

I think this is the cleanest way to get around the problem that the parser doesn't know how to interpret text attributes that occur _earlier_ than the `CA` tag. What do you think, @yishn?

**Comment by yishn** [comment]
Thank you so much for your work @apetresc! I'll look over it once I'm back.

**Comment by yishn** [comment]
Yes, it's probably the best way. If the CA tag only appears in the root node, we only need to parse just a few tokens to get to the CA property.

**Comment by yishn** [comment]
Another idea is as follows, since the CA property can only appear in the root node: With the current implementation, the only properties not handled are the properties in the parent node of the CA property. A quick (and maybe dirty) way is to just fix said properties when we encounter a CA tag. This would also work if there are several games with different CA properties inside one sgf file.

**Comment by yishn** [comment]
I'm back from vacation! @apetresc do you want to finish this PR, or should I take over?

**Comment by apetresc** [comment]
I can finish it! I'll implement your suggestion on Monday.

How do you feel about the saving aspect? Do you want to preserve the `CA[]` tag and encode with it when saving, or convert everything to `UTF-8`?

**Comment by yishn** [comment]
I don't really know, but personally, I'm fine with saving everything in UTF-8.

**Comment by apetresc** [comment]
@yishn, I think I'm done! I've completed all the objectives set out initially, and it works well from my personal manual testing, including saving back to UTF-8.

Let me know if there's anything else that needs doing :)

File: modules/sgf.js [hunk 1]
```diff
@@ -10,11 +10,21 @@ if (typeof require != 'undefined') {
...
 }
 
 var context = typeof module != 'undefined' ? module.exports : (window.sgf = {})
 var alpha = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
 
+/* The default encoding and list of properties that should be interpreted as
+ * being encoded by the file's CA[] property is defined in the SGF spec at
+ * http://www.red-bean.com/sgf/properties.html#CA
+ */
+var default_encoding = 'ISO-8859-1'
```
Comment (yishn) [inline modules/sgf.js]: Can you change the variable names to camelCase style to fit with convention?

File: package.json [hunk 2]
```diff
@@ -16,9 +16,11 @@
   "dependencies": {
     "argv-split": "^1.0.0",
     "gemini-scrollbar": "^1.4.3",
+    "iconv-lite": "^0.4.13",
     "octicons": "^4.3.0",
     "pikaday": "^1.4.0",
-    "sigma": "^1.0.3"
+    "sigma": "^1.0.3",
+    "tmp": "0.0.28"
```
Comment (yishn) [inline package.json]: Shouldn't `tmp` be in devDependencies if it's only needed in the tests?

**Comment by yishn** [comment]
Thanks for your hard work! I left some comments in the code. I have one question, though: 

In [case we have an encoding and try to parse values](https://github.com/yishn/Sabaki/pull/122/files#diff-7183a0b6f43c8f1ab8ceef19d6bb450cR113) we first decode the encoded data, then apply unescapeString, i.e. `unescapeString(decode(encoded_data))`, however when we [encounter `CA` and reparse values](https://github.com/yishn/Sabaki/pull/122/files#diff-7183a0b6f43c8f1ab8ceef19d6bb450cR98), we actually do `decode(unescapeString(encoded_data))`. I'm wondering if this yields the same thing?

I have no experience with text encodings and I'm not sure what the "correct" order is, but they seem to give the same results.

File: modules/sgf.js (hunk 1 above)
Comment (apetresc) [inline modules/sgf.js]: Whoops, not sure how I missed that! Done :)

File: package.json (hunk 2 above)
Comment (apetresc) [inline package.json]: Definitely – I just didn't know `devDependencies` existed, this is my first time touching Node. Done!

**Comment by apetresc** [comment]
Excellent point about the different order of escaping. In theory, it *could* make a difference if the encoding we're switching between had a different byte for `"` or `\` characters. In practice, though, virtually all 8-bit encodings in common use today (and probably all the ones that `iconv-lite` actually supports) are ASCII-compatible and don't differ on those common characters.

**Comment by yishn** [comment]
Thank you for clarifying! I'm merging this now.

**Pull request merged**